0_functions/simulation_functions.R

```diff
@@ -261,3 +261,78 @@ betabinom_cdf <- function(x, size, p, phi) rmutil::pbetabinom(x, size, p, phi)
 binom_pmf <- function(x, size, p) dbinom(x, size, p)
 betabinom_pmf <- function(x, size, p, phi) rmutil::dbetabinom(x, size, p, phi)
 
+
+
+# load and modify the BBC pandemic data as we need to for the draw_gatherings() function below
+bbc <- read_csv("1_data/contact_dist_BBCPandemic/contact_distributions_o18.csv")
+bbc <- select(bbc, e_other)
+# gathering sizes are contacts + 1 (index person)
+bbc$M <- bbc$e_other + 1
+# plotting it
+bbc_data <- ggplot(data = bbc, mapping = aes(x = M)) + 
+  geom_point(stat = "count") + 
+  scale_y_log10() + scale_x_log10()
+pdf("3_results/bbc_data.pdf", width = 5, height = 6)
+print(bbc_data)
+dev.off()  
+
+# draw_gatherings() function generates M, a vector of gatherings drawn from the
+# BBC pandemic distribution so that in total N people attend,
+# under 4 options for restriction in gathering size up to c : 0 for no restriction, 
+# 1 for replacement by c, 2 for replacement by 0 and 3 for replacement by another draw smaller to c
+
+draw_gatherings <- function(N = 10000, #10000 as default, can change
+                            c = 10, # cut-off as default, can change
+                            option) { # define options for restriction (1, 2, or 3)
+# initialize loop variables
+M <- vector()
+i <- 1
+# create subset of bbc distribution <= c (for option 3)
+smallM <- sample(bbc$M[bbc$M <= c])
+# loop until sum of gatherings is equal to pop size
+while(sum(M) != N) {
+  # draw gathering of size m from BBC pandemic distribution
+  M[i] <- sample(bbc$M, 1) 
+  if (M[i] > c) {
+    M[i] <- switch(
+      option,
+      # option 0 : no restriction to gathering size 
+      "0" = M[i],
+      # option 1: set to c (most conservative)
+      "1" = c,
+      # option 2: set to 1 (least conservative)
+      "2" = 1,
+      # option 3: redraw until less than c (realistic?) 
+      #"3" = while (M[i] > c) # Chris I'm not sure what you were trying to do here (and it didn't seem to work?)
+      #  sample(bbc$M, 1)
+      "3" = sample(smallM, 1) #So I did that instead
+    )
+  }
+  # if sum(m) is less than pop size, then advance to next gathering
+  # otherwise redraw existing gathering
+  if (sum(M) <= N) { 
+    i <- i + 1
+  } 
+} 
+return(M)
+} 
+
+# apply the draw_SIR_attendees() function to each of the gatherings from draw_gatherings()
+# get distribution of S, I and R in each gathering 
+
+gath <- as.matrix(draw_gatherings(option = "0" ))
+head(gath)
+test1 <- lapply(gath,draw_SIR_attendees,ps= 0.95, pi = 0.05, sims = 1)
+head(test1)
+test2 <- matrix(unlist(test1), ncol = 3, byrow = TRUE)
+head(test2)
+test3 <- data.frame(gath, test2)
+colnames(test3) <- c("N", "S", "I", "R")
+head(test3)
+#sum(test3$N) # check yes
```

Comment:
This code works, but I was thinking that we would do something more like:

```
N <- 10000
pi <- 0.05
pr <- 0.05 

# draw gatherings
gath <- draw_gatherings(option = "0")

# define vector of population states based on pi and pr
states <- c(
  rep("S", N - floor(pi * N) - floor(pr * N)),
  rep("I", floor(pi * N)),
  rep("R", floor(pr * N))
)

# for each gathering sample M from vector of population states to attend
# gathering and tabulate numbers of S, I, and R
gath_states <- lapply(gath, function (x) table(sample(states, x))) 
gath_states <- do.call('rbind', gath_states)

# combine into final data.frame
gath <- cbind("N" = gath, gath_states)
```

This way we have fixed number of S, I, and R in population and are just randomly assigning them to gatherings as opposed to using multinomial distribution in which case we may end up with proportion of infecteds across all gatherings could be > or < pi just due to sampling variability.